pkg/net/rpc.go

```diff
@@ -1,66 +1,46 @@
 package net
 
 import (
+	"bytes"
 	"context"
 	"encoding/binary"
-	"io/ioutil"
-	"time"
-
-	"github.com/libp2p/go-libp2p-core/discovery"
 	"github.com/libp2p/go-libp2p-core/helpers"
 	"github.com/libp2p/go-libp2p-core/host"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/libp2p/go-libp2p-core/protocol"
 	"github.com/libp2p/go-libp2p-core/record"
 	protoutil "github.com/wetware/casm/pkg/util/proto"
+	"golang.org/x/sync/errgroup"
+	"math/rand"
+	"sort"
 )
 
 /*
- * This file contains the implementation for the random walk that subtends Overlay.FindPeers.
- *
- * The algorithm is remarkably simple, even though the unholy mix of distribution and concurrency
- * makes its implementation quite verbose.  A random walk works like this:
- *
- *  1. FindPeers selects a depth parameter, corresponding to the number of steps in the random walk.  The default is 9.
- *  2. FindPeers selects a neighbor at random, opens a stream to the /casm/net/sample endpoint, and transmits depth-1.
- *  3. The remote peer receives the depth parameter.  If depth == 1, it replies with the address of one of its neighbors at random.†
- *  4. Else depth > 1, and the remote peer picks up the random walk at step #2 (recursion).
- *
- * That's it. Everything else is glue code.
- *
- * When reading over this code, you should skip over anything you don't understand.  Don't shave the proverbial yak.
- *
- * † A quick explanation, in case 'depth == 1' surprised you.  We random walk until 1 instead of 0 to avoid an extra round-trip
- *   on the network. When we get to the penultimate node, we need only pick one of its neighbors at random and return its listen
- *   address.  We do not need to open a stream to it.
+ * TODO: Write Gossiping short explanation?
  */
 
 const (
 	ProtocolVersion             = "0.0.0"
 	BaseProto       protocol.ID = "/casm/net"
-
-	defaultSampleLimit       = 1
-	defaultSampleDepth       = 7
-	defaultSampleStepTimeout = time.Second * 30
 )
 
 var (
-	// sampleProto and joinProto are not used for routing.  Their only
+	// joinProto is not used for routing.  The only
 	// purpose is to serve as a key when (un)registering protocol handlers.
 	//
 	// They should never appear on the wire.
 	joinProto   = protoutil.Join(BaseProto, "join")
-	sampleProto = protoutil.Join(BaseProto, "sample")
+	gossipProto = protoutil.Join(BaseProto, "gossip")
 
-	versionProto protocol.ID = protoutil.Join(BaseProto, protocol.ID(ProtocolVersion))
+	versionProto = protoutil.Join(BaseProto, ProtocolVersion)
```

Comment:
Here I cleaned up the code because we could omit explicitly stating `protocol.ID()`

Reply:
Nice, thanks for catching that!